File: worker/lib/notification/trigger.ex
# Copyright (C) 2021-2023 tiksan
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

defmodule Tornium.Trigger do
  alias Tornium.Repo
  import Ecto.Query

  @api_call_priority 10

  @spec execute(trigger :: Tornium.Schema.Trigger.t(), api_key :: Tornium.Schema.TornKey.t()) ::
          {:ok, List} | {:error, any()} | {:lua_error, any()}
  def execute(%Tornium.Schema.Trigger{} = trigger, %Tornium.Schema.TornKey{} = api_key)
      when trigger.resource == :user do
    # TODO: Document this function
    update_next_execution(trigger)

    response =
      Tornex.Scheduler.Bucket.enqueue(%Tornex.Query{
        resource: "user",
        resource_id: trigger.resource_id,
        selections: trigger.selections,
        key: api_key.api_key,
        key_owner: api_key.user_id,
        nice: @api_call_priority
      })

    case response do
      %{"error" => %{"code" => code, "error" => error}} ->
        {:error, Tornium.API.Error.construct(code, error)}

      %{} ->
        # TODO: Handle errors from `Tornium.Lua.execute_lua`
        Tornium.Lua.execute_lua(trigger.code, user: response)
    end
  end

  def execute(%Tornium.Schema.Trigger{} = _trigger) do
    {:error, :not_implemented}
  end

  @spec parse_next_execution(trigger :: Tornium.Schema.Trigger.t()) :: DateTime.t()
  defp parse_next_execution(%Tornium.Schema.Trigger{} = trigger) do
    # TODO: Document this function
    Crontab.CronExpression.Parser.parse!(trigger.cron)
    |> Crontab.Scheduler.get_next_run_date!()
    |> DateTime.from_naive!("Etc/UTC")
  end

  @spec update_next_execution(trigger :: Tornium.Schema.Trigger.t()) :: nil
  def update_next_execution(%Tornium.Schema.Trigger{} = trigger) do
    # TODO: Document this function
    {1, _} =
      Tornium.Schema.Trigger
      |> where(tid: ^trigger.tid)
      |> update(set: [next_execution: ^parse_next_execution(trigger.cron)])
      |> Repo.update_all([])

    nil
  end
end

File: application/static/notification/trigger.js
/* Copyright (C) 2021-2023 tiksan

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as published
by the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>. */

window.triggerErrorCallback = function(jsonError, container) {
    const errorNode = document.createElement("span");
    errorNode.textContent = " Data failed to load...";

    const errorIcon = document.createElement("i");
    errorIcon.classList.add("fa-solid", "fa-skull-crossbones");

    container.innerHTML = "";
    container.appendChild(errorIcon);
    container.appendChild(errorNode);
}

window.addTriggerToViewer = function(trigger, triggerContainer) {
    const triggerNode = document.createElement("div");
    triggerNode.classList.add("card", "mx-2", "mt-2", "viewer-card");
    triggerNode.setAttribute("data-trigger-id", trigger.tid);
    triggerContainer.append(triggerNode);

    const triggerRow = document.createElement("div");
    triggerRow.classList.add("row", "p-2", "align-middle");
    triggerNode.append(triggerRow);

    const triggerNameElement = document.createElement("div");
    triggerNameElement.classList.add("col-sm-12", "col-md-2", "col-xl-2");
    triggerNameElement.textContent = trigger.name;
    triggerRow.append(triggerNameElement);

    const triggerDescriptionElement = document.createElement("div");
    triggerDescriptionElement.classList.add("col-sm-12", "col-md-8", "col-xl-6", "text-truncate");
    triggerDescriptionElement.textContent = trigger.description;
    triggerRow.append(triggerDescriptionElement);

    const triggerPaddingElement = document.createElement("div");
    triggerPaddingElement.classList.add("col-sm-0", "col-md-0", "col-xl-2");
    triggerRow.append(triggerPaddingElement);

    const triggerActionsContainer = document.createElement("div");
    triggerActionsContainer.classList.add("col-sm-12", "col-md-2", "col-xl-2");
    triggerRow.append(triggerActionsContainer);
    
    const triggerActions = document.createElement("div");
    triggerActions.classList.add("w-100", "justify-content-end", "d-flex");
    triggerActionsContainer.append(triggerActions);

    const viewAction = document.createElement("a")
    viewAction.classList.add("btn", "btn-sm", "btn-outline-secondary", "me-2");
    viewAction.href = `/notification/trigger/view/${trigger.tid}`;
    triggerActions.append(viewAction);

    const viewActionIcon = document.createElement("i");
    viewActionIcon.classList.add("fa-regular", "fa-eye");
    viewAction.append(viewActionIcon);
}
